cert/tls.go

```diff
@@ -58,3 +101,50 @@ func TLSConfFromCert(certData tls.Certificate) *tls.Config {
 		MinVersion:   tls.VersionTLS12,
 	}
 }
+
+// TLSReloader updates the TLS certificate without restarting the server.
+type TLSReloader struct {
+	certMu sync.RWMutex
+	cert   *tls.Certificate
+}
+
+// NewTLSReloader is used to create a new TLS Reloader that will be used
+// to update the TLS certificate without restarting the server.
+func NewTLSReloader(certBytes, keyBytes []byte) (*TLSReloader, error) {
+	cert, _, err := LoadCertFromBytes(certBytes, keyBytes)
+	if err != nil {
+		return nil, err
+	}
+	return &TLSReloader{
+		cert: &cert,
+	}, nil
+}
+
+// AttemptReload will make an attempt to update the TLS certificate
+// and key used by the server.
+func (t *TLSReloader) AttemptReload(certBytes, keyBytes []byte) error {
+	newCert, _, err := LoadCertFromBytes(certBytes, keyBytes)
+	if err != nil {
+		return err
+	}
+
+	t.certMu.Lock()
+	t.cert = &newCert
+	defer t.certMu.Unlock()
```

Comment:
There are two patterns with locks that we commonly use to help readability.
Either:

```
func (s *struct) myMethod() {
	s.mu.Lock()
	defer s.mu.Unlock()

	// Do something here while knowing the lock is being held for the whole time. Notice the empty line after the defer to make this stand out better.
```

Or:

```
func (s *struct) myMethod() {
	// Do something that doesn't require the lock being held to minimize the time other threads could be blocked.

	// Do the one thing that requires the lock to be held, this 3-line pattern _without_ the defer usually stands out pretty well.
	s.mu.Lock()
	s.unsafeVar = foo
	s.mu.Unlock()

	// Do other things that don't require holding the lock.
```

So tl;dr: please remove the `defer` here and add a newline between the `defer` and `return` in the method below.

Reply:
Got it :+1: